File: App/Bracket48Core/Sources/Bracket48Core/Scoring.swift
public struct ScoringRuleSet: Equatable, Sendable {
    public let correctGroupWinner: Int
    public let correctGroupRunnerUp: Int
    public let correctGroupThirdPlace: Int
    public let correctThirdPlaceAdvancement: Int
    public let perfectGroupTopThreeBonus: Int
    public let knockoutRoundPoints: [KnockoutRound: Int]

    public init(
        correctGroupWinner: Int,
        correctGroupRunnerUp: Int,
        correctGroupThirdPlace: Int,
        correctThirdPlaceAdvancement: Int,
        perfectGroupTopThreeBonus: Int,
        knockoutRoundPoints: [KnockoutRound: Int]
    ) {
        self.correctGroupWinner = correctGroupWinner
        self.correctGroupRunnerUp = correctGroupRunnerUp
        self.correctGroupThirdPlace = correctGroupThirdPlace
        self.correctThirdPlaceAdvancement = correctThirdPlaceAdvancement
        self.perfectGroupTopThreeBonus = perfectGroupTopThreeBonus
        self.knockoutRoundPoints = knockoutRoundPoints
    }

    public static let worldCupDefault = ScoringRuleSet(
        correctGroupWinner: 4,
        correctGroupRunnerUp: 3,
        correctGroupThirdPlace: 2,
        correctThirdPlaceAdvancement: 2,
        perfectGroupTopThreeBonus: 3,
        knockoutRoundPoints: [
            .roundOf32: 4,
            .roundOf16: 6,
            .quarterfinal: 8,
            .semifinal: 12,
            .final: 20
        ]
    )
}

public enum ScoreSourceType: String, Codable, Sendable {
    case groupStagePrediction
    case knockoutPick
}

public struct ScoreEvent: Equatable, Sendable {
    public let id: String
    public let entryID: TournamentEntryID
    public let sourceType: ScoreSourceType
    public let sourceID: String
    public let points: Int
    public let reason: String

    public init(
        id: String,
        entryID: TournamentEntryID,
        sourceType: ScoreSourceType,
        sourceID: String,
        points: Int,
        reason: String
    ) {
        self.id = id
        self.entryID = entryID
        self.sourceType = sourceType
        self.sourceID = sourceID
        self.points = points
        self.reason = reason
    }
}

public struct ScoreBreakdown: Equatable, Sendable {
    public let entryID: TournamentEntryID
    public let groupStagePoints: Int
    public let knockoutPoints: Int
    public let events: [ScoreEvent]

    public var totalPoints: Int {
        groupStagePoints + knockoutPoints
    }

    public init(
        entryID: TournamentEntryID,
        groupStagePoints: Int,
        knockoutPoints: Int,
        events: [ScoreEvent]
    ) {
        self.entryID = entryID
        self.groupStagePoints = groupStagePoints
        self.knockoutPoints = knockoutPoints
        self.events = events
    }
}

public struct MaximumScore: Equatable, Sendable {
    public let groupStagePoints: Int
    public let knockoutPoints: Int

    public var totalPoints: Int {
        groupStagePoints + knockoutPoints
    }
}

public struct ScoringEngine: Sendable {
    public let rules: ScoringRuleSet

    public init(rules: ScoringRuleSet = .worldCupDefault) {
        self.rules = rules
    }

    public func score(
        entryID: TournamentEntryID,
        predictions: BracketPredictions,
        results: TournamentResults
    ) -> ScoreBreakdown {
        let groupStandingsByID = Dictionary(
            uniqueKeysWithValues: results.groupStandings.map { ($0.groupID, $0) }
        )
        let knockoutResultsByMatchID = Dictionary(
            uniqueKeysWithValues: results.knockoutResults.map { ($0.matchID, $0) }
        )

        var groupStageEvents: [ScoreEvent] = []
        for prediction in predictions.groupStagePredictions.sorted(by: { $0.groupID.rawValue < $1.groupID.rawValue }) {
            guard let standing = groupStandingsByID[prediction.groupID] else {
                continue
            }

            groupStageEvents.append(
                contentsOf: scoreGroupStagePrediction(
                    entryID: entryID,
                    prediction: prediction,
                    standing: standing,
                    advancingThirdPlaceTeamIDs: results.advancingThirdPlaceTeamIDs
                )
            )
        }

        var knockoutEvents: [ScoreEvent] = []
        for pick in predictions.knockoutPicks.sorted(by: knockoutPickSort) {
            guard let result = knockoutResultsByMatchID[pick.matchID],
                  result.winnerTeamID == pick.pickedWinnerTeamID
            else {
                continue
            }

            let points = rules.knockoutRoundPoints[pick.round, default: 0]
            knockoutEvents.append(
                ScoreEvent(
                    id: "knockout:\(pick.matchID.rawValue):winner",
                    entryID: entryID,
                    sourceType: .knockoutPick,
                    sourceID: pick.matchID.rawValue,
                    points: points,
                    reason: "Correct \(pick.round.displayName) winner"
                )
            )
        }

        let events = groupStageEvents + knockoutEvents
        return ScoreBreakdown(
            entryID: entryID,
            groupStagePoints: groupStageEvents.reduce(0) { $0 + $1.points },
            knockoutPoints: knockoutEvents.reduce(0) { $0 + $1.points },
            events: events
        )
    }

    public func maximumAvailableScore(for predictions: BracketPredictions) -> MaximumScore {
        let groupPointsPerPrediction = rules.correctGroupWinner
            + rules.correctGroupRunnerUp
            + rules.correctGroupThirdPlace
            + rules.correctThirdPlaceAdvancement
            + rules.perfectGroupTopThreeBonus

        let groupStagePoints = predictions.groupStagePredictions.count * groupPointsPerPrediction
        let knockoutPoints = predictions.knockoutPicks.reduce(0) { total, pick in
            total + rules.knockoutRoundPoints[pick.round, default: 0]
        }

        return MaximumScore(groupStagePoints: groupStagePoints, knockoutPoints: knockoutPoints)
    }

    private func scoreGroupStagePrediction(
        entryID: TournamentEntryID,
        prediction: GroupStagePrediction,
        standing: FinalGroupStanding,
        advancingThirdPlaceTeamIDs: Set<TeamID>
    ) -> [ScoreEvent] {
        var events: [ScoreEvent] = []
        let predicted = prediction.orderedTeamIDs
        let actual = standing.orderedTeamIDs

        if predicted[safe: 0] == actual[safe: 0] {
            events.append(
                groupStageEvent(
                    entryID: entryID,
                    groupID: prediction.groupID,
                    ruleID: "winner",
                    points: rules.correctGroupWinner,
                    reason: "Correct group winner"
                )
            )
        }

        if predicted[safe: 1] == actual[safe: 1] {
            events.append(
                groupStageEvent(
                    entryID: entryID,
                    groupID: prediction.groupID,
                    ruleID: "runner-up",
                    points: rules.correctGroupRunnerUp,
                    reason: "Correct group runner-up"
                )
            )
        }

        if predicted[safe: 2] == actual[safe: 2] {
            events.append(
                groupStageEvent(
                    entryID: entryID,
                    groupID: prediction.groupID,
                    ruleID: "third-place",
                    points: rules.correctGroupThirdPlace,
                    reason: "Correct third-place team"
                )
            )
        }

        if let actualThirdPlace = actual[safe: 2],
           prediction.predictedThirdPlaceAdvances == advancingThirdPlaceTeamIDs.contains(actualThirdPlace)
        {
            events.append(
                groupStageEvent(
                    entryID: entryID,
                    groupID: prediction.groupID,
                    ruleID: "third-place-advancement",
                    points: rules.correctThirdPlaceAdvancement,
                    reason: "Correct third-place advancement result"
                )
            )
        }

        if Array(predicted.prefix(3)) == Array(actual.prefix(3)) {
            events.append(
                groupStageEvent(
                    entryID: entryID,
                    groupID: prediction.groupID,
                    ruleID: "perfect-top-three",
                    points: rules.perfectGroupTopThreeBonus,
                    reason: "Perfect group top 3"
                )
            )
        }

        return events.sorted { lhs, rhs in
            if lhs.sourceID == rhs.sourceID {
                return lhs.id < rhs.id
            }

            return lhs.sourceID < rhs.sourceID
        }
    }

    private func groupStageEvent(
        entryID: TournamentEntryID,
        groupID: GroupID,
        ruleID: String,
        points: Int,
        reason: String
    ) -> ScoreEvent {
        ScoreEvent(
            id: "group:\(groupID.rawValue):\(ruleID)",
            entryID: entryID,
            sourceType: .groupStagePrediction,
            sourceID: groupID.rawValue,
            points: points,
            reason: reason
        )
    }

    private func knockoutPickSort(_ lhs: KnockoutPick, _ rhs: KnockoutPick) -> Bool {
        if lhs.round.sortOrder == rhs.round.sortOrder {
            return lhs.matchID.rawValue < rhs.matchID.rawValue
        }

        return lhs.round.sortOrder < rhs.round.sortOrder
    }
}

private extension KnockoutRound {
    var sortOrder: Int {
        switch self {
        case .roundOf32:
            1
        case .roundOf16:
            2
        case .quarterfinal:
            3
        case .semifinal:
            4
        case .final:
            5
        }
    }

    var displayName: String {
        switch self {
        case .roundOf32:
            "Round of 32"
        case .roundOf16:
            "Round of 16"
        case .quarterfinal:
            "quarterfinal"
        case .semifinal:
            "semifinal"
        case .final:
            "final"
        }
    }
}

private extension Array {
    subscript(safe index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: Backend/supabase/functions/_shared/scoring.ts
export const scoringVersion = "world-cup-default-v1";

const rules = {
  correctGroupWinner: 4,
  correctGroupRunnerUp: 3,
  correctGroupThirdPlace: 2,
  correctThirdPlaceAdvancement: 2,
  perfectGroupTopThreeBonus: 3,
  knockoutRoundPoints: {
    roundOf32: 4,
    roundOf16: 6,
    quarterfinal: 8,
    semifinal: 12,
    final: 20,
  } as Record<string, number>,
};

export type BracketPhase = "group_stage" | "knockout";

export type PoolEntryRow = {
  id: string;
  pool_id: string;
  bracket_id: string;
  user_id: string;
  phase: BracketPhase;
  brackets?: {
    id: string;
    phase: BracketPhase;
    picks: BracketPicks;
  };
};

export type BracketPicks = {
  predictions?: GroupPrediction[];
  picks?: KnockoutPick[];
};

export type GroupPrediction = {
  group_id: string;
  ordered_team_ids: string[];
  predicted_third_place_advances: boolean;
};

export type KnockoutPick = {
  match_id: string;
  round: string;
  picked_winner_team_id: string;
};

export type GroupStandingResult = {
  group_id: string;
  ordered_team_ids: string[];
};

export type KnockoutResult = {
  match_id: string;
  round: string;
  winner_team_id: string;
  eliminated_team_ids?: string[];
};

export type ScoreEvent = {
  source_type: "group_stage_prediction" | "knockout_pick";
  source_id: string;
  rule_id: string;
  points: number;
  reason: string;
};

export type ScoreBreakdown = {
  pool_entry_id: string;
  pool_id: string | null;
  bracket_id: string | null;
  user_id: string | null;
  phase: BracketPhase | "combined";
  group_stage_points: number;
  knockout_points: number;
  total_points: number;
  max_points: number;
  possible_points_remaining: number;
  events: ScoreEvent[];
};

export type ScoringResults = {
  groupStandings: GroupStandingResult[];
  finalGroupIDs: Set<string>;
  advancingThirdPlaceTeamIDs: Set<string>;
  knockoutResults: KnockoutResult[];
  eliminatedTeamIDs: Set<string>;
};

export type DatabaseGroupStanding = {
  group_name: string;
  teams: Array<{
    team_id: string | null;
    position: number;
    points: number;
    played?: number | null;
    goal_difference: number | null;
    goals_for: number | null;
  }>;
};

export function scorePoolEntry(
  entry: PoolEntryRow,
  results: ScoringResults,
): ScoreBreakdown {
  return scorePicks({
    poolEntryID: entry.id,
    poolID: entry.pool_id,
    bracketID: entry.bracket_id,
    userID: entry.user_id,
    phase: entry.phase,
    picks: entry.brackets?.picks ?? {},
    results,
  });
}

export function scoreSimulationEntries(
  entries: Array<Record<string, unknown>>,
  results: ScoringResults,
): ScoreBreakdown[] {
  return entries.map((entry, index) => {
    const phase = entry["phase"] === "knockout" || entry["phase"] === "group_stage" ? entry["phase"] : "combined";
    return scorePicks({
      poolEntryID: typeof entry["entry_id"] === "string" ? entry["entry_id"] : `simulation-${index + 1}`,
      poolID: null,
      bracketID: null,
      userID: typeof entry["user_id"] === "string" ? entry["user_id"] : null,
      phase,
      picks: {
        predictions: arrayValue(entry["predictions"]) as GroupPrediction[],
        picks: arrayValue(entry["knockout_picks"]) as KnockoutPick[],
      },
      results,
    });
  });
}

export function scorePicks(input: {
  poolEntryID: string;
  poolID: string | null;
  bracketID: string | null;
  userID: string | null;
  phase: BracketPhase | "combined";
  picks: BracketPicks;
  results: ScoringResults;
}): ScoreBreakdown {
  const events: ScoreEvent[] = [];
  const groupStandingsByID = new Map(input.results.groupStandings.map((standing) => [standing.group_id, standing]));
  const knockoutResultsByMatchID = new Map(input.results.knockoutResults.map((result) => [result.match_id, result]));

  const groupPredictionScores = new Map<string, number>();
  for (const prediction of input.picks.predictions ?? []) {
    const standing = groupStandingsByID.get(prediction.group_id);
    if (!standing) {
      continue;
    }

    const groupEvents = scoreGroupPrediction(prediction, standing, input.results.advancingThirdPlaceTeamIDs);
    groupPredictionScores.set(
      prediction.group_id,
      groupEvents.reduce((total, event) => total + event.points, 0),
    );
    events.push(...groupEvents);
  }

  for (const pick of input.picks.picks ?? []) {
    const result = knockoutResultsByMatchID.get(pick.match_id);
    if (!result || result.winner_team_id !== pick.picked_winner_team_id) {
      continue;
    }

    const points = rules.knockoutRoundPoints[pick.round] ?? 0;
    events.push({
      source_type: "knockout_pick",
      source_id: pick.match_id,
      rule_id: "winner",
      points,
      reason: `Correct ${roundDisplayName(pick.round)} winner`,
    });
  }

  const groupStagePoints = events
    .filter((event) => event.source_type === "group_stage_prediction")
    .reduce((total, event) => total + event.points, 0);
  const knockoutPoints = events
    .filter((event) => event.source_type === "knockout_pick")
    .reduce((total, event) => total + event.points, 0);

  return {
    pool_entry_id: input.poolEntryID,
    pool_id: input.poolID,
    bracket_id: input.bracketID,
    user_id: input.userID,
    phase: input.phase,
    group_stage_points: groupStagePoints,
    knockout_points: knockoutPoints,
    total_points: groupStagePoints + knockoutPoints,
    max_points: maximumScore(input.picks),
    possible_points_remaining: possiblePointsRemaining(input.picks, input.results, groupPredictionScores),
    events,
  };
}

export function scoreGroupPrediction(
  prediction: GroupPrediction,
  standing: GroupStandingResult,
  advancingThirdPlaceTeamIDs: Set<string>,
): ScoreEvent[] {
  const events: ScoreEvent[] = [];
  const predicted = prediction.ordered_team_ids;
  const actual = standing.ordered_team_ids;

  if (predicted[0] === actual[0]) {
    events.push(groupEvent(prediction.group_id, "winner", rules.correctGroupWinner, "Correct group winner"));
  }

  if (predicted[1] === actual[1]) {
    events.push(groupEvent(prediction.group_id, "runner-up", rules.correctGroupRunnerUp, "Correct group runner-up"));
  }

  if (predicted[2] === actual[2]) {
    events.push(groupEvent(prediction.group_id, "third-place", rules.correctGroupThirdPlace, "Correct third-place team"));
  }

  const actualThirdPlaceTeamID = actual[2];
  if (
    actualThirdPlaceTeamID
    && prediction.predicted_third_place_advances === advancingThirdPlaceTeamIDs.has(actualThirdPlaceTeamID)
  ) {
    events.push(groupEvent(
      prediction.group_id,
      "third-place-advancement",
      rules.correctThirdPlaceAdvancement,
      "Correct third-place advancement result",
    ));
  }

  if (predicted.slice(0, 3).join("|") === actual.slice(0, 3).join("|")) {
    events.push(groupEvent(prediction.group_id, "perfect-top-three", rules.perfectGroupTopThreeBonus, "Perfect group top 3"));
  }

  return events;
}

export function maximumScore(picks: BracketPicks): number {
  const groupPointsPerPrediction = rules.correctGroupWinner
    + rules.correctGroupRunnerUp
    + rules.correctGroupThirdPlace
    + rules.correctThirdPlaceAdvancement
    + rules.perfectGroupTopThreeBonus;
  const groupStagePoints = (picks.predictions ?? []).length * groupPointsPerPrediction;
  const knockoutPoints = (picks.picks ?? []).reduce((total, pick) => total + (rules.knockoutRoundPoints[pick.round] ?? 0), 0);
  return groupStagePoints + knockoutPoints;
}

export function possiblePointsRemaining(
  picks: BracketPicks,
  results: ScoringResults,
  currentGroupPredictionScores = new Map<string, number>(),
): number {
  const groupPointsPerPrediction = rules.correctGroupWinner
    + rules.correctGroupRunnerUp
    + rules.correctGroupThirdPlace
    + rules.correctThirdPlaceAdvancement
    + rules.perfectGroupTopThreeBonus;
  const knockoutResultsByMatchID = new Set(results.knockoutResults.map((result) => result.match_id));

  const groupStageRemaining = (picks.predictions ?? []).reduce((total, prediction) => {
    if (results.finalGroupIDs.has(prediction.group_id)) {
      return total;
    }

    const provisionalPoints = currentGroupPredictionScores.get(prediction.group_id) ?? 0;
    return total + Math.max(0, groupPointsPerPrediction - provisionalPoints);
  }, 0);

  const knockoutRemaining = (picks.picks ?? []).reduce((total, pick) => {
    if (knockoutResultsByMatchID.has(pick.match_id)) {
      return total;
    }

    if (results.eliminatedTeamIDs.has(pick.picked_winner_team_id)) {
      return total;
    }

    return total + (rules.knockoutRoundPoints[pick.round] ?? 0);
  }, 0);

  return groupStageRemaining + knockoutRemaining;
}

export function finalGroupIDs(groups: DatabaseGroupStanding[]): Set<string> {
  return new Set(groups
    .filter((group) => group.teams.length >= 4 && group.teams.every((team) => (team.played ?? 0) >= 3))
    .map((group) => groupIDFromName(group.group_name)));
}

export function advancingThirdPlaceTeams(groups: DatabaseGroupStanding[]): Set<string> {
  return new Set(groups
    .flatMap((group) => group.teams.map((team) => ({ ...team, group_name: group.group_name })))
    .filter((team) => team.position === 3 && !!team.team_id)
    .sort((lhs, rhs) =>
      rhs.points - lhs.points
      || (rhs.goal_difference ?? 0) - (lhs.goal_difference ?? 0)
      || (rhs.goals_for ?? 0) - (lhs.goals_for ?? 0)
      || lhs.group_name.localeCompare(rhs.group_name)
    )
    .slice(0, 8)
    .map((team) => team.team_id!));
}

export function parseGroupStanding(value: unknown): GroupStandingResult | null {
  const object = objectValue(value);
  if (!object || typeof object["group_id"] !== "string") {
    return null;
  }

  const orderedTeamIDs = arrayValue(object["ordered_team_ids"])
    .filter((teamID): teamID is string => typeof teamID === "string");
  return { group_id: object["group_id"], ordered_team_ids: orderedTeamIDs };
}

export function parseKnockoutResult(value: unknown): KnockoutResult | null {
  const object = objectValue(value);
  if (
    !object
    || typeof object["match_id"] !== "string"
    || typeof object["round"] !== "string"
    || typeof object["winner_team_id"] !== "string"
  ) {
    return null;
  }

  return {
    match_id: object["match_id"],
    round: object["round"],
    winner_team_id: object["winner_team_id"],
    eliminated_team_ids: arrayValue(object["eliminated_team_ids"])
      .filter((teamID): teamID is string => typeof teamID === "string"),
  };
}

export function internalKnockoutMatchID(round: string, index: number): string {
  switch (round) {
  case "roundOf32":
    return `r32-${index}`;
  case "roundOf16":
    return `r16-${index}`;
  case "quarterfinal":
    return `qf-${index}`;
  case "semifinal":
    return `sf-${index}`;
  case "final":
    return "final";
  default:
    return `${round}-${index}`;
  }
}

export function appRound(databaseRound: string): string | null {
  switch (databaseRound) {
  case "round_of_32":
    return "roundOf32";
  case "round_of_16":
    return "roundOf16";
  case "quarterfinal":
  case "semifinal":
  case "final":
    return databaseRound;
  default:
    return null;
  }
}

export function groupIDFromName(groupName: string): string {
  return groupName.replace(/^Group\s+/i, "").trim();
}

export function numberOrNull(value: unknown): number | null {
  return typeof value === "number" ? value : null;
}

export function objectValue(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : null;
}

export function arrayValue(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

export async function sha256(value: string): Promise<string> {
  const bytes = new TextEncoder().encode(value);
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return [...new Uint8Array(digest)].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

function groupEvent(groupID: string, ruleID: string, points: number, reason: string): ScoreEvent {
  return {
    source_type: "group_stage_prediction",
    source_id: groupID,
    rule_id: ruleID,
    points,
    reason,
  };
}

function roundDisplayName(round: string): string {
  switch (round) {
  case "roundOf32":
    return "Round of 32";
  case "roundOf16":
    return "Round of 16";
  case "quarterfinal":
    return "Quarterfinal";
  case "semifinal":
    return "Semifinal";
  case "final":
    return "Final";
  default:
    return "knockout";
  }
}
